src/driver-openinput.c

```diff
@@ -0,0 +1,483 @@
+{
+	int ret;
+	struct openinput_drv_data *drv_data = ratbag_get_drv_data(device);
+	struct oi_report_t report = {
+		.id = OI_REPORT_SHORT,
+		.function_page = OI_PAGE_INFO,
+		.function = OI_FUNCTION_FW_INFO,
+		.data = {field_id}
+	};
+
+	ret = openinput_send_report(device, &report);
+	if (ret)
+		return ret;
+
+	strncpy(description, report.data, min(sizeof(report.data), description_size));
+
+	return 0;
+}
+
+
+static int openinput_info_supported_function_pages(struct ratbag_device *device,
+						   uint8_t start_index,
+						   uint8_t *pages_count,
+						   uint8_t *pages_left,
+						   uint8_t *pages,
+						   size_t pages_size)
+{
+	int ret;
+	struct oi_report_t report = {
+		.id = OI_REPORT_SHORT,
+		.function_page = OI_PAGE_INFO,
+		.function = OI_FUNCTION_SUPPORTED_FUNCTION_PAGES,
+		.data = {start_index}
+	};
+
+	ret = openinput_send_report(device, &report);
+	if (ret)
+		return ret;
+
+	*pages_count = report.data[0];
+	*pages_left = report.data[1];
+	memcpy(pages, report.data + 2, min(sizeof(report.data), pages_size));
+
+	return 0;
+}
+
+static int openinput_info_supported_functions(struct ratbag_device *device,
+					      uint8_t function_page,
+					      uint8_t start_index,
+					      uint8_t *functions_count,
+					      uint8_t *functions_left,
+					      uint8_t *functions,
+					      size_t functions_size)
+{
+	int ret;
+	struct oi_report_t report = {
+		.id = OI_REPORT_SHORT,
+		.function_page = OI_PAGE_INFO,
+		.function = OI_FUNCTION_SUPPORTED_FUNCTIONS,
+		.data = {function_page, start_index}
+	};
+
+	ret = openinput_send_report(device, &report);
+	if (ret)
+		return ret;
+
+	*functions_count = report.data[0];
+	*functions_left = report.data[1];
+	memcpy(functions, report.data + 2, min(sizeof(report.data), functions_size));
+
+	return 0;
+}
+
+static int openinput_read_supported_functions(struct ratbag_device *device, uint8_t page)
+{
+	struct ratbag *ratbag = device->ratbag;
+	struct openinput_drv_data *drv_data = ratbag_get_drv_data(device);
+	int ret;
+	uint8_t i, total, read = 0, count = 0, left = 0;
+	uint8_t buffer[OI_REPORT_DATA_MAX_SIZE];
+
+	ret = openinput_info_supported_functions(device,
+						 page,
+						 read,
+						 &count, &left,
+						 buffer, sizeof(buffer));
+	if (ret)
+		return ret;
+
+	total = count + left;
+	uint8_t functions[total];
+
+	memcpy(functions, buffer, count);
+
+	/* there are still functions left to read! */
+	while (left) {
+		ret = openinput_info_supported_functions(device,
+							 page,
+							 read,
+							 &count, &left,
+							 buffer, sizeof(buffer));
+		if (ret)
+			return ret;
+
+		/* make sure the new size values make sense, to avoid deadlocks */
+		if (total != (read + count + left))
+		{
+			log_error(ratbag, "openinput: invalid number of functions left to read (%u)\n", left);
+			return -EINVAL;
+		}
+		log_debug(ratbag, "openinput: read %u functions, %u left\n", count, left);
+
+		memcpy(functions + read, buffer, count);
+		read += count;
+	}
```

Comment:
very much a nitpick but it *seems* like this could be a `do { } while()` instead to not have two separate but identical invocations of `openinput_info_supported_functions`. haven't fully engaged my brain to make sure that's true though.

Reply:
I did that originally but it was a bit of a mess doing to the validation logic, note the declaration of `functions` based on a value return by the command, so I opted to duplicate the command making it a bit easier to read. This does seem like a good optimization at first sight but I found that the code duplication is much easier to read than the increased logic.